File: 686/Main.java
import java.io.*;
import java.util.*;

public class Main {
   public static void main(String args[]) throws IOException {
      boolean[] primeTable = new boolean[32768];
      String line;
      StringBuilder answer = new StringBuilder();
      ArrayList<Integer> prime = new ArrayList<Integer>();
      BufferedReader in = new BufferedReader(new InputStreamReader(System.in));

      primeTable[2] = true;
      for (int i = 3; i < primeTable.length; i += 2)
         primeTable[i] = true;

      for (int i = 3; i * i < 32768; i += 2) {
         for (int j = i; i * j < 32768; j += 2) {
            primeTable[i * j] = false;
         }
      }

      prime.add(2);
      for (int i = 3; i < 32768; i += 2) {
         if (primeTable[i]) prime.add(i);
      }

      while ((line = in.readLine()) != null) {
         int n = Integer.parseInt(line);
         int count = 0;
         if (n == 0) break;
         for (int i = 0; i < prime.size(); i++) {
            int p = prime.get(i);
            if ((p * 2) > n) break;
            count += primeTable[(n - p)] ? 1 : 0;
         }
         answer.append(Integer.toString(count));
         answer.append('\n');
      }
      System.out.print(answer);
   }
}

File: 686/main.c
#include <stdio.h>
#include <memory.h>

int main(int argc, char* argv[])
{
   int i, j, k, count;
   int prime[4000];
   int prime_len = 0;
   char prime_table[32768];

   for (i = 0; i < 32768; i++)
      prime_table[i] = (i % 2 != 0);

   for (i = 3; i * i < 32768; i += 2) {
      for (j = i, k = i * i; k < 32768; j += 2, k = i * j) {
         prime_table[k] = 0;
      }
   }

   prime_table[2] = 1;
   prime[prime_len++] = 2;
   for (i = 3; i < 32768; i += 2)
      if (prime_table[i]) prime[prime_len++] = i;

   while (scanf("%d", &k) == 1 && k != 0) {
      count = 0;
      for (i = 0; i < prime_len; i++) {
         if ((prime[i] * 2) > k) break;
         count += prime_table[(k - prime[i])];
      }
      printf("%d\n", count);
   }

   return 0;
}
